### ISSUE TITLE
using alpha monitor the training process of the model

### ISSUE THREAD

**Issue by cobraheleah** [CalculatedContent/WeightWatcher#307]
I use alpha value to monitor the training effect of the model, but I found that as the training progresses, the alpha value will rapidly decrease in the early stage and slowly increase in the middle and later stages. The specific data is as follows:

training step, alpha value
1k, 28.94
2k, 15
3k, 7.1
5k, 4.4
6k, 4.2
100k, 4.7
200k, 4.9
300k, 5.0
400k, 5.1
500k, 5.2

However, throughout the entire training process, the evaluation effect of the model (eg. mmlu, bbh) has been improving.
Is it normal for the alpha value to slowly increase in the middle and later stages of training? 
It seems that the size of the alpha value is not absolutely related to the evaluation effect of the model.

**Comment by charlesmartin14** [comment]
how many layers does your model have ?how are you computing the average alpha ?Sent from my iPhoneOn Mar 6, 2024, at 6:51 PM, cobraheleah ***@***.***> wrote:﻿
I use alpha value to monitor the training effect of the model, but I found that as the training progresses, the alpha value will rapidly decrease in the early stage and slowly increase in the middle and later stages. The specific data is as follows:
training step, alpha value
1k, 28.94
2k, 15
3k, 7.1
5k, 4.4
6k, 4.2
100k, 4.7
200k, 4.9
300k, 5.0
400k, 5.1
500k, 5.2
However, throughout the entire training process, the evaluation effect of the model (eg. mmlu, bbh) has been improving.
Is it normal for the alpha value to slowly increase in the middle and later stages of training?
It seems that the size of the alpha value is not absolutely related to the evaluation effect of the model.

**Comment by charlesmartin14** [comment]
1) I suspect Llama is too big for your data set.
In fact, Llama itself is not well size...see this comparison with Falcon 
https://weightwatcher.ai/leaderboard.html

Both Falcon and Mistral show much better quality scores that LLama

it appears that as some of the LLama layers are learning information, others are just becoming more random.
There are 2 things you can do

1a) check for Dragon Kings
https://calculatedcontent.com/2024/01/29/evaluating-llms-with-weightwatcher-part-iii-the-magic-of-mistral-a-story-of-dragon-kings/

1b) don't include any layer with alpha > 8 in your average

2) Use robust statistics...that is, compute the median alpha and/or throw away outliers.
2a) I can add method to weightwatcher to do this 

2b) Also, If you are training models from scratch, then spectral density of the data jacobian / gradients is also useful to observe.
But this is very expensive to compute,  and usually  cost prohibitive

3) The alpha range comes from the JMLR theory paper
https://jmlr.org/papers/v22/20-410.html


![HTSR](https://github.com/CalculatedContent/WeightWatcher/assets/498448/9034bb4c-dcdf-4c34-992c-431dbfe9bd31)

The theory is an infinite theory limit, so the values should not depend on the size and shape of W
But in practice, it does depend on these a little bit, so I usually say 6 is a good upper limit in general



4*) If you are training Llama from scatch, you might want to try the GaLore optimizer

Memory-Efficient LLM Training by Gradient Low-Rank Projection
https://lnkd.in/g2P_HTPE

**Comment by charlesmartin14** [comment]
>  I don’t think it is due to insufficient data. ...
> , Other open-source models like Baichuan2-7B which Published the intermediate ckpt also exhibit this phenomenon that the alpha value first decreases and then increases. 

Like Llama, the  Baichuan2-7B model is thought to have underfit  / redundant layers.   
See the recent ShortGPT paper
https://arxiv.org/abs/2403.03853

Because these layers are not converging and /or are redudant, the model is not well sized, and its possible that other layers are 'soaking up' the correlations, causing their layers alphas to be smaller than expected

---

 If you are going to apply weightwatcher early in training  you need to check a few things because is quite possible that the fits early in training are simply spurious because the layer is so far from convergence

  You can fit any data set and get an alpha,   so in addition to computing alpha, you have to check that

- the tail is large enough to get a reliable fit

- that the quality of fit (D)

- the PL fit is stable 

- the ESD is unimodal, heavy-tailed, and sufficiently different from random

- there are no Correlation Traps which can cause spuriously small alphas

- there is no rank collapse 

- the layer alphas for the model correlate well with other metrics, such as the rand_distance, spectral norm, distance from init
 (see this blog post: https://calculatedcontent.com/2021/10/17/fantastic-measures-of-generalization-that-actually-work-part-1/_)

- the eigenvectors of the tail have lower entropy than the bulk

> the alpha value of 40b-instruct Falcon is higher than 65b Llama, is
The quality of the alpha fit is frequently a more reliable metric than the value of alpha itself
![Screenshot 2024-03-12 at 9 13 24 AM](https://github.com/CalculatedContent/WeightWatcher/assets/498448/f482b6cc-834d-4bd2-a710-729849405a46)



> A better approach would be to compare the distribution of alpha values?
weightwatcher is a diagnostic tool for analyzing how models converge, layer-by-layer
It can be used to study how  individual layer converges, the correlation flow, how layers inter-correlates with each other ,etc
